File: app/src/main/java/de/openfiresource/falarm/models/Rest.java
package de.openfiresource.falarm.models;

import javax.inject.Inject;
import javax.inject.Singleton;

import de.openfiresource.falarm.models.api.OpenPagerService;
import de.openfiresource.falarm.models.api.UserKey;
import de.openfiresource.falarm.models.api.UserLogin;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import timber.log.Timber;

@Singleton
public class Rest {

    private final OpenPagerService service;

    @Inject
    Rest() {
        HttpLoggingInterceptor logging = new HttpLoggingInterceptor(message -> Timber.tag("OkHttp").v(message));
        logging.setLevel(HttpLoggingInterceptor.Level.BODY);


        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(logging)
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("http://area51.openpager.de/api/v1/")
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

        service = retrofit.create(OpenPagerService.class);
    }

    Single<UserKey> login(UserLogin userLogin) {
        return service.login(userLogin).subscribeOn(Schedulers.io());
    }

    public Single<ResponseBody> logout() {
        return service.logout().subscribeOn(Schedulers.io());
    }
}

File: app/src/main/java/de/openfiresource/openpager/models/Rest.kt
package de.openfiresource.openpager.models

import de.openfiresource.openpager.models.api.*
import de.openfiresource.openpager.utils.Constants
import de.openfiresource.openpager.utils.Preferences
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class Rest @Inject internal constructor(preferences: Preferences) {

    private val service: OpenPagerService

    init {
        val logging = HttpLoggingInterceptor { message -> Timber.tag("OkHttp").v(message) }
        logging.level = HttpLoggingInterceptor.Level.BODY

        val client = OkHttpClient.Builder()
                .addInterceptor(TokenInterceptor(preferences))
                .addInterceptor(logging)
                .build()

        val retrofit = Retrofit.Builder()
                .baseUrl(Constants.BACKEND_URL_API)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()

        service = retrofit.create(OpenPagerService::class.java)
    }

    internal fun login(userLogin: UserLogin): Single<UserKey> {
        return service.login(userLogin).subscribeOn(Schedulers.io())
    }

    internal fun logout(): Completable {
        return service.logout().subscribeOn(Schedulers.io())
    }

    internal fun putDeviceInfo(device: Device): Completable {
        return service.putDeviceInfo(device.fcmToken, device).subscribeOn(Schedulers.io())
    }
}
